### ISSUE TITLE
Review reporting indicators to assess feasibility of creating a child list from reporting

### ISSUE THREAD

**Issue by cafootitt** [OpenSRP/opensrp-client-chw#162]
Reporting indicator definitions are here: https://docs.google.com/spreadsheets/d/1q9YiWqjLiToTd0--Q8CbwhBwwcNDspbDxVrUfDm8VGU/edit#gid=1627225414

Please review and assess the feasibility of whether we can create a child list to match the filters of the indicators for when the end user taps on a slice of the pie chart, e.g. to see something like this:

<img width="209" alt="screen shot 2019-02-27 at 9 43 23 am" src="https://user-images.githubusercontent.com/3488986/53470961-49c86d00-3a74-11e9-819d-d7411aa7ab10.png">

**Comment by cagulas** [comment]
Spoke to the client, and I explained how this is a bit complex. He says we can de-prioritize this until we work on the ANC module later this year. 

@keymane I can move this to the icebox column for now. I don't believe there would be anything we do differently in the implementation of the in-app reporting module that we'll need to take into consideration now, as this seems to be add-on functionality, right?

**Comment by cafootitt** [comment]
@ellykits @allan-allay Can you both review this issue and jot down the implementation plan for it and add a t-shirt size? I'm going to need to get this resourced ahead of the Togo roll-out.

**Comment by ellykits** [comment]
**Technical implementation**

From a technical perspective, this issue can either be addressed in the library or the functionality can as well be added to the projects that are using the reporting library. What this would take is overriding the functionality of the view responsible for creating the pie chart and appending this functionality in the pie chart listener exposed by the library.

This functionality can still be part of the library as suggested but can be made configurable in a way that implementing projects can still define their own custom implementation. 

In a nutshell this is what the implementation would involve: 

1. Define queries that would return the data to be displayed on the view. Each query mapped to pie chart slice. 
2. Create the view for displaying the information when a pie chart is clicked (Currently on clicking pie chart slice displays a toaster)
3. Override the ```ChartListener ```class and implement the functionality that will run the queries and display them in the view created in step 2 above

**Comment by craigappl** [comment]
Ultimately, we would want to make this configurable across any client in the system so that the developer can build any report in the system and return any list of clients. The larger requirement is to make any report be able to query any of the client tables in the SQLite database, return a list of results that are dynamic and configurable in the view and use that as a navigation element to explore the client record.

For example, this would also be a really cool feature in the Daily Tally sheet, where a user could tap a particular value and get a view of children who were seen. They would then need to tap one of the children to view the profile with breadcrumbs returning to the original query.

I think we will need to create a roadmap for this feature that builds in maturity over time. From the project perspective, it could look something like this:
**MVP**
- Lay the foundation for the bigger picture building a generic hook (I don't know the Android term here) that can be called from any report with the parameters passed in to generate the query
- Build a generic dynamic view that would display the list of patients (or, we could improve the filters in the register view and just display that with a back button)
**Iteration 1**
- Improve the view so that they are configurable like we have in the configurable views feature where columns are returned based on a file in the system
- Improve the navigation so people can drill all the way down into the record of the entity (be it a person, family, etc) and find their way back using the core views that we already have in the system
**Iteration 2** (I'm not sure if we need this)
- Open up the ability to chain or cascade these reports so that one report can open another report that's more granular which will ultimately get down to the list of people

**Comment by cafootitt** [comment]
**Notes from 2 August call:**

- Need to involve Roger on design to show the filter selected on the register view. Need to have some cancel mechanism if the user wants to cancel the process and land on the full register list.
- Pagination should be handled already in the register list
- Need to add tag to the indicators to know which register view to land on
- What do we need to pass into the register view (ideally just patient IDs)
- Pass IDs and the indicator name into the register list.
- Do we run the query again in the register view to refresh the filtered list of children, so that children who are actioned aren't further included in the view.
- Back button should refresh reporting when you go back to the reporting view?

**Comment by rowo** [comment]
Recent decision: I'm going to do some light thinking about it, but no hands on design work yet until @mberg the product owner is completely on board. The next step isn't getting me involved unless it's a tricky requirements design problem — it should be an internal presentation from @craigappl and/or @cafootitt about to him about the problem and pros/cons of the chosen solution.

General advice on involving product owners:
1. a new feature (hasn’t been done before)
2. involves an ongoing product approach (not just a one off project feature)
3. unclear about the need to do it now

**Comment by cafootitt** [comment]
@mberg At the risk of me wrongly summarizing the outcome of our discussion last night, would you be able to provide your final determination here?

**Comment by cafootitt** [comment]
I discussed with @mberg and we can go ahead and mock this up and design the interaction. In terms of priority to implement it, it would be after the PNC deadlines we have in August for Wcaro and BA.

**Comment by cafootitt** [comment]
@rowo Thanks for this. About your side note, I think your description doesn't match the mockup. You're proposing keeping the "Yes" and "No" labels, but the mockup provided shows the "not up to date" phrase. 

If you're saying we should use "Yes" and "No", can you provide a mockup for the "No" version?

**Comment by cafootitt** [comment]
@rowo Ok thanks. An updated mockup would be really helpful for clarity when I create a github implementation issue for our team to implement, since it's easier to just follow the mockup.

**Comment by ellykits** [comment]
> @ellykits What was your estimate for that initial proposal of using a generic dynamic view?

@rowo From our previous discussion with @allan-allay  we estimated the issue to be **medium-size**. However with the proposed implementation we are yet to give an estimate that is after further discussion; putting into consideration the filters that will now be applied to the respective registers instead. We shall have a meeting some time in the course of the week preferably this Friday to talk about this then give back the feedback.

**Comment by githengi** [comment]
We choose to use the registers to generate option 1.

**Defining indicators to have additional attributes ( <=2 day)**
- Register  
- Id column  (id column is used in getting the ids)

**Reporting Library Changes**
Tagging slice label with the indicator **<= 1 day**
Executing the indicator query to get the Ids  and pass to register **<= 1 days**
Opening the register and passing the required params **<=1 day**


**Register customization**
Making the toolbar configurable  **<=1 day**
Passing Ids from reporting library as filters to the fts query **<=1 day**
Only displaying the patient demographics details as in option **1 <= 2 days**

**Testing** 
Unit testing  **<= 1 day**

**Contingency**
2 days

WCaro Integration for existing child and ANC indicators ~ **2 days**


**Optional** 
Support of charts derived using grouped indicators **<= 5 days**

**Comment by cafootitt** [comment]
Total: 19 days, but we can deliver this for wcaro without the grouped indicators feature, so 3 weeks for wcaro delivery.

**Comment by rowo** [comment]
@githengi If there was any confusion... the rows in the list in https://github.com/OpenSRP/opensrp-client-chw/issues/162#issuecomment-526231328 are tappable to see the profiles

**Comment by cafootitt** [comment]
Client prefers the simplified view with Approach 1 for naming the indicators. Closing this issue and creating a new issue for implementation.
